### PR TITLE
Ensure BaseVector::copy only accepts rows valid in source

### LINKED ISSUE #6591: Do not allow invalid values in toSourceRow argument of BaseVector::copy
### Bug description

PR #4255 modified BaseVector::copy to allow invalid values in 'toSourceRow'. The change was to silently ignore these. This is problematic as it can hide difficult to track down bugs. We need to remove this check.

(Note that the code was moved from .h to .cpp file in 5989.)

```
void BaseVector::copy(
    const BaseVector* source,
    const SelectivityVector& rows,
    const vector_size_t* toSourceRow) {
  // Check if there are rows that do not exist in 'source'. Remove these from
  // 'ranges'.
  // TODO Update the callers and remove this logic.

  std::vector<CopyRange> ranges;
  if (toSourceRow == nullptr) {
    ...
  } else {
    ranges.reserve(rows.end());
    rows.applyToSelected([&](vector_size_t row) {
      const auto sourceRow = toSourceRow[row];

      // TODO Remove this check or turn it into DCHECK.
      if (sourceRow >= source->size()) {
        return;
      }
      ranges.push_back({sourceRow, row, 1});
    });
  }

  if (ranges.empty()) {
    return;
  }

  copyRanges(source, ranges);
}
```

CC: @laithsakka @bikramSingh91 @Yuhta @oerling 

### System information

n/a

### Relevant logs

_No response_

### PR DESCRIPTION
**Pull request by bikramSingh91**
Fixes #6591
Found 2 bugs so far:
Addressed in #6967 and #6968
Depends on those fixes for a clean run of unit tests and fuzzer

Differential Revision: D49341013

### CODE REVIEW

**Review by Yuhta — COMMENTED** [review]
Can you also fix the one in `FlatVector-inl.h`?

**Comment by bikramSingh91** [comment]
Looking at the failure in OperatorUtilsTest.projectChildren.
It seems like NestedLoopJoinProbe implementation has logic that create a RowVector first and then replaces its children. This might not work correctly due to bug 2 mentioned in the description because if the children are not replaced (in case not all children are projected) then it can result in wrong results. I will look into what the right fix here should be. Ideally, we should not directly replace children inside a RowVector by using its childAt() API since we can end up replacing it with a child that renders the RowVector malformed.

**Comment by kevinwilfong** [comment]
Can you add unit tests to illustrate what bugs you fixed and ensure they don't come back?

Also, is it necessary to combine these 2 fixes into a single diff?  It'd be easier to review as separate changes.

**Comment by bikramSingh91** [comment]
Looks like the fuzzer caught another bug by the check added in this PR, will look into it: https://app.circleci.com/pipelines/github/facebookincubator/velox/34465/workflows/9b61c6d1-bff3-40ac-be53-0affebeaa5a2/jobs/225273

**Comment by bikramSingh91** [comment]
Tired the fuzzer failure after rebase, seems to have fixed it, probably was a transient issue due to previous rebase which is not fixed.

**Comment by bikramSingh91** [comment]
Separated out Fixes for bugs found so far in #6967 and https://github.com/facebookincubator/velox/pull/6968

File: velox/vector/BaseVector.cpp [hunk 1]
```diff
@@ -669,36 +669,21 @@ void BaseVector::copy(
...
       ranges = toCopyRanges(rows);
-    }
   } else {
     ranges.reserve(rows.end());
     rows.applyToSelected([&](vector_size_t row) {
       const auto sourceRow = toSourceRow[row];
-      if (sourceRow >= source->size()) {
-        return;
-      }
+      VELOX_CHECK_GT(source->size(), sourceRow);
```
Comment (Yuhta) [inline velox/vector/BaseVector.cpp:683]: `VELOX_DCHECK_GT`?

**Comment by bikramSingh91** [comment]
the latest fuzzer failure in CI is an unrelated known issue: #7126

File: velox/vector/BaseVector.cpp (hunk 1 above)
Comment (bikramSingh91) [reply velox/vector/BaseVector.cpp:683]: @Yuhta I added DCHECKS only for the paths that are executed per row, for all other they are executed once per vector so should not affect perf.
Please let me know if you still prefer I change it to DCHECK

Comment (Yuhta) [reply velox/vector/BaseVector.cpp:683]: This particular line is per row I think?

Comment (bikramSingh91) [reply velox/vector/BaseVector.cpp:683]: aha, good catch! didnt realize it from the tiny code snippet here. Thanks for reviewing!

**Review by Yuhta — APPROVED** [review]
There are some lint warnings

**Pull request closed without merging**